anchor-remote/src/cas_remote.rs

```diff
@@ -0,0 +1,330 @@
+        let jws_header = Header {
+            kid: format!(
+                "{}#{}",
+                controller,
+                controller
+                    .strip_prefix("did:key:")
+                    .expect("invalid did:key")
+            ),
+            alg: "EdDSA".to_string(),
+            typ: "JWT".to_string(),
+        };
+        let jws_header_b64 =
+            b64.encode(serde_json::to_vec(&jws_header).expect("invalid jws header"));
+        let node_stream_id = cid_to_stream_id(cid_from_ed25519_key_pair(&keypair));
+        Self {
+            signing_key: keypair,
+            url: format!("{}/api/v0/requests", remote_anchor_service_url),
+            poll_interval: anchor_poll_interval,
+            poll_retry_count: anchor_poll_retry_count,
+            jws_header_b64,
+            http_client: reqwest::Client::new(),
+            node_stream_id,
+        }
+    }
+
+    /// Create an anchor request on the remote CAS
+    pub async fn create_anchor_request(&self, root_cid: Cid) -> Result<String> {
+        let cas_request_body = serde_json::to_string(&CasAnchorRequest {
+            stream_id: self.node_stream_id.clone(),
+            cid: root_cid.to_string(),
+            timestamp: chrono::Utc::now().to_rfc3339(),
+            ceramic_one_version: AGENT_VERSION.to_owned(),
+        })?;
+        let digest = MultihashDigest::digest(&Code::Sha2_256, cas_request_body.as_bytes());
+        let digest = hex::encode(digest.digest());
+        let auth_jwt = self.auth_jwt(format!("0x{}", digest)).await?;
+        let auth_header = format!("Bearer {}", auth_jwt);
+        debug!("auth_header {}", &auth_header);
+        let res = self
+            .http_client
+            .post(self.url.clone())
+            .header("Authorization", auth_header)
+            .header("Content-Type", "application/json")
+            .body(cas_request_body)
+            .send()
+            .await?;
+        Ok(res.text().await?)
+    }
+
+    async fn auth_jwt(&self, digest: String) -> Result<String> {
+        let body = Claims {
+            digest,
+            nonce: Uuid::new_v4().to_string(),
+            url: self.url.clone(),
+        };
+        let body_b64 = b64.encode(serde_json::to_vec(&body)?);
+        let message = [self.jws_header_b64.clone(), body_b64].join(".");
+        let sig_b64 = b64.encode(self.signing_key.sign(message.as_bytes()));
+        Ok([message.clone(), sig_b64].join("."))
+    }
+}
+
+async fn parse_anchor_response(anchor_response: String) -> Result<CasResponseParseResult> {
+    // Return if we were unable to parse the anchor response
+    let anchor_response = serde_json::from_str::<CasAnchorResponse>(anchor_response.as_str())?;
+
+    // If the response does not contain a witness CAR file, the anchor request is either still pending or it failed
+    // because the request was unauthorized.
+    let Some(witness_car_b64) = anchor_response.witness_car else {
```

Comment:
fancy.. I had to look up how long `let/else` has been stable (since 1.65 and I've never used it! 🙊)